src/main/java/com/cleanroommc/groovyscript/compat/mods/prodigytech/AtomicReshaper.java

```diff
@@ -0,0 +1,166 @@
+    }
+
+    @Property(property = "input", valid = @Comp("1"))
+    @Property(property = "output", valid = @Comp("1"))
+    public static class RecipeBuilder extends AbstractRecipeBuilder<AtomicReshaperManager.AtomicReshaperRecipe> {
+
+        @Property(valid = @Comp(value = "1", type = Comp.Type.GTE))
+        private int time = -1;
+
+        @Property(valid = @Comp(value = "1", type = Comp.Type.GTE))
+        private int primordium;
+
+        @Property(valid = @Comp("1"))
+        private List<Object> weightedOutputs;
+
+        RecipeBuilder() {
+            weightedOutputs = new ArrayList<>();
+        }
+
+        private String errorMessage = "";
+
+        @RecipeBuilderMethodDescription
+        public AtomicReshaper.RecipeBuilder time(int time) {
+            this.time = time;
+            return this;
+        }
+
+        @RecipeBuilderMethodDescription
+        public AtomicReshaper.RecipeBuilder primordium(int primordium) {
+            this.primordium = primordium;
+            return this;
+        }
+
+        @RecipeBuilderMethodDescription
+        public AtomicReshaper.RecipeBuilder addOutput(ItemStack output, int weight) {
+            if (weight <= 0 && errorMessage.isEmpty()) {
+                errorMessage = String.format("ItemStack %s has invalid weight %d, expected >= 1", output, weight);
+            } else {
+                weightedOutputs.add(output);
+                weightedOutputs.add(weight);
+            }
+            return this;
+        }
+
+        @Override
+        public String getErrorMsg() {
+            return "Error adding ProdigyTech Atomic Reshaper Recipe";
+        }
+
+        @Override
+        public void validate(GroovyLog.Msg msg) {
+            validateItems(msg, 1, 1, 0, 1);
+            validateFluids(msg);
+
+            boolean hasNormalOutputs = !output.isEmpty();
+            boolean hasWeightedOutputs = !weightedOutputs.isEmpty();
+            msg.add(hasNormalOutputs && hasWeightedOutputs, "Both normal and weighted outputs were provided!");
+            msg.add(!hasNormalOutputs && !hasWeightedOutputs, "Neither of normal and weighted outputs were provided!");
+
+            msg.add(!errorMessage.isEmpty(), "Weighted outputs error: {}", errorMessage);
+            msg.add(primordium <= 0, "primordium must be greater than or equal to 1, yet it was {}", primordium);
+            // 100 is hardcoded in the source
+            int capacity = Config.atomicReshaperMaxPrimordium * 100;
+            msg.add(primordium > capacity, "primordium must be less than or equal to the Reshaper's capacity {}, yet it was {}", capacity, primordium);
+            if (time <= 0) {
+                time = Config.atomicReshaperProcessTime;
+            }
+        }
+
+        private Object[] getRecipeOutput() {
+            if (!output.isEmpty()) {
+                return new ItemStack[]{output.get(0)};
+            } else {
+                return weightedOutputs.toArray();
+            }
+        }

```

Comment:
the way `weightedOutputs` works is nasty. i would suggest replacing the `weightedOutputs` Object List with a `chance` Integer List. then, do this
```suggestion
        private Object[] getRecipeOutput() {
            List<Object> target = new ArrayList<>();
            for (int i = 0; i < output.size(); i++) {
                target.add(output.get(i));
                target.add(chance.get(i));
            }
            return target.toArray();
        }
```
this would also significantly simplify validation, since now you just have to allow `>=1` outputs and check that `chance.size()` and `output.size()` are equal.

Reply:
notably there is a separate `output("one guaranteed output")` and also `output("my output", weight)`, but i guess I'll do this regardless, I do agree this implementation is nasty